core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java

```diff
@@ -615,20 +678,45 @@ private void sendUpdate()
-        Optional<PlanFragment> fragment = sendPlan.get() ? Optional.of(planFragment.withoutEmbeddedJsonRepresentation()) : Optional.empty();
-        TaskUpdateRequest updateRequest = new TaskUpdateRequest(
-                session.toSessionRepresentation(),
-                session.getIdentity().getExtraCredentials(),
-                fragment,
-                splitAssignments,
-                outputBuffers.get(),
-                dynamicFilterDomains.getDynamicFilterDomains(),
-                session.getExchangeEncryptionKey());
-        byte[] taskUpdateRequestJson = taskUpdateRequestCodec.toJsonBytes(updateRequest);
+        List<SplitAssignment> splitAssignments;
+        VersionedDynamicFilterDomains dynamicFilterDomains;
+        Optional<PlanFragment> fragment;
+        byte[] taskUpdateRequestJson;
+        synchronized (this) {
+            int oldSplitBatchSize = splitBatchSize;
+            splitAssignments = getSplitAssignments();
+            dynamicFilterDomains = outboundDynamicFiltersCollector.acknowledgeAndGetNewDomains(sentDynamicFiltersVersion.get());
+
+            // Workers don't need the embedded JSON representation when the fragment is sent
+            fragment = sendPlan.get() ? Optional.of(planFragment.withoutEmbeddedJsonRepresentation()) : Optional.empty();
+            TaskUpdateRequest updateRequest = new TaskUpdateRequest(
+                    session.toSessionRepresentation(),
+                    session.getIdentity().getExtraCredentials(),
+                    fragment,
+                    splitAssignments,
+                    outputBuffers.get(),
+                    dynamicFilterDomains.getDynamicFilterDomains(),
+                    session.getExchangeEncryptionKey());
+            taskUpdateRequestJson = taskUpdateRequestCodec.toJsonBytes(updateRequest);
+
+            if (enableAdaptiveTaskRequestSize) {
+                // try to adjust batch size to meet expected request size: (requestSizeLimit - requestSizeLimitHeadroom)
+                int newSplitBatchSize = adjustSplitBatchSize(splitAssignments, taskUpdateRequestJson.length, maxRequestSize - requestSizeHeadroom, oldSplitBatchSize);
+                // abandon current request and reschedule update if size of request body exceeds requestSizeLimit
+                // and splitBatchSize is updated
+                if (newSplitBatchSize != oldSplitBatchSize) {
+                    splitBatchSize = newSplitBatchSize;
+                    splitBatchSizeChanged = true;
+                }
+                if (taskUpdateRequestJson.length > maxRequestSize && newSplitBatchSize < oldSplitBatchSize) {
+                    log.info("%s - current taskUpdateRequestJson exceeded limit: %d, oldSplitBatchSize: %d, adjustedSplitBatchSize: %d, reschedule.",
+                            taskId, taskUpdateRequestJson.length, oldSplitBatchSize, newSplitBatchSize);
+                    scheduleUpdate();
```

Comment:
Shouldn't we test here `taskUpdateRequestJson.length > (maxRequestSize - requestSizeHeadroom)`?

Reply:
If `requestSize = 6 mb` and `maxRequestSize = 7 mb`, since `requestSize < maxRequestSize`, it doesn't need to reschedule. `requestSizeHeadroom` is for calculating the number of `splits` only when `requestSize > maxRequestSize` (given that a request contains not only splits, but also other data). Let me change the logic to call `adjustSplitBatchSize` only when `requestSize > maxRequestSize`.